dnf-json

```diff
@@ -123,98 +126,147 @@ def repo_checksums(base):
 
     return checksums
 
-
-call = json.load(sys.stdin)
-command = call["command"]
-arguments = call["arguments"]
-repos = arguments.get("repos", {})
-arch = arguments["arch"]
-cachedir = arguments["cachedir"]
-module_platform_id = arguments["module_platform_id"]
-
-with tempfile.TemporaryDirectory() as persistdir:
-    try:
-        base = create_base(
-            repos,
-            module_platform_id,
-            persistdir,
-            cachedir,
-            arch
-        )
-    except dnf.exceptions.Error as e:
-        exit_with_dnf_error(
-            type(e).__name__,
-            f"Error occurred when setting up repo: {e}"
-        )
-
-    if command == "dump":
-        packages = []
-        for package in base.sack.query().available():
-            packages.append({
-                "name": package.name,
-                "summary": package.summary,
-                "description": package.description,
-                "url": package.url,
-                "epoch": package.epoch,
-                "version": package.version,
-                "release": package.release,
-                "arch": package.arch,
-                "buildtime": timestamp_to_rfc3339(package.buildtime),
-                "license": package.license
-            })
-        json.dump({
-            "checksums": repo_checksums(base),
-            "packages": packages
-        }, sys.stdout)
-
-    elif command == "depsolve":
-        errors = []
-
-        try:
-            base.install_specs(
-                arguments["package-specs"],
-                exclude=arguments.get("exclude-specs", [])
-            )
-        except dnf.exceptions.MarkingErrors as e:
-            exit_with_dnf_error(
-                "MarkingErrors",
-                f"Error occurred when marking packages for installation: {e}"
-            )
-
-        try:
-            base.resolve()
-        except dnf.exceptions.DepsolveError as e:
-            exit_with_dnf_error(
-                "DepsolveError",
-                (
-                    "There was a problem depsolving "
-                    f"{arguments['package-specs']}: {e}"
+class DnfJsonRequestHandler(BaseHTTPRequestHandler):
+
+    def _send(self):
+        self.client_address=('',)
+
+    def exit_with_dnf_error(self, kind: str, reason: str):
+        self._send()
+        self.send_response(500)
```

Comment:
This preserves the behaviour (I think), so it is the right thing for the first version. But I would love to see this improved to return structured 400 errors in case of user error to explain all the different ways something might have gone wrong (conflicting requests, invalid repository, ...).

Reply:
Yes it does preserve the behaviour because we check in composer the value of the HTTP code. Before we were checking the value of the return code of DNF.

At the moment we already forward the reason why to the end user:

```
 time curl -k --key /etc/osbuild-composer/client-key.pem --cert /etc/osbuild-composer/client-crt.pem -H "Content-Type: application/json" -d @../tests/1477.json https://localhost/api/composer/v1/compose -i 
HTTP/1.1 400 Bad Request
Content-Type: application/json; charset=UTF-8
Date: Wed, 24 Nov 2021 17:02:46 GMT
Content-Length: 225

{"message":"Failed to depsolve requested package set: DNF error occured: MarkingErrors: Error occurred when marking packages for installation: Problems in request:\nmissing packages: jesuisunpaquetquinexistepas_idonotexist"}

real	0m0.754s
user	0m0.033s
sys	0m0.009s
```

Reply:
I think this is right for now, but in the future we should make this a structured object so we can show the error message nicely in the UI. And we should distinguish between 400 (user made a mistake) and 500 (we made a mistake). Something like:
```
{
  "type": "MissingPackage",
  "payload": {
    "packages": [
      "jesuisunpaquetquinexistepas_idonotexist"
    ]
  }
}
```

Reply:
Since depsolve is a job now  and asynchronous (as in, the request succeeds if the job was queued). So it wouldn't result in 400s/500s. But of course we still need a good way of distinguishing errors yea.

Reply:
Ok so I was not testing it with the v2. Now I'm doing it and I can see that indeed there's no way for a user to get feedback, at least through composer's api.

```
time curl -k --key /etc/osbuild-composer/client-key.pem --cert /etc/osbuild-composer/client-crt.pem -H "Content-Type: application/json" -d @../tests/compose_request_v2.json https://localhost/api/image-builder-composer/v2/compose -i 
HTTP/1.1 201 Created
Content-Type: application/json; charset=UTF-8
Date: Thu, 25 Nov 2021 14:44:54 GMT
Content-Length: 113

{"href":"/api/image-builder-composer/v2/compose","kind":"ComposeId","id":"2989cd1f-5647-4783-ae58-6a2770cf89e1"}

real	0m0.043s
user	0m0.031s
sys	0m0.005s
┌─(✓)─[tlavocat@fedora]─[16% <─ ~/dev/tests]
└──╼> time curl -k --key /etc/osbuild-composer/client-key.pem --cert /etc/osbuild-composer/client-crt.pem -H "Content-Type: application/json" -d @../tests/compose_request_v2.json https://localhost/api/image-builder-composer/v2/composes/2989cd1f-5647-4783-ae58-6a2770cf89e1 -i 
HTTP/1.1 404 Not Found
Content-Type: application/json; charset=UTF-8
Date: Thu, 25 Nov 2021 14:52:49 GMT
Content-Length: 201

{"href":"/api/image-builder-composer/v2/errors/21","id":"21","kind":"Error","code":"IMAGE-BUILDER-COMPOSER-21","operation_id":"21PoEE2eQFJMusCbAfSNVy9rXKv","reason":"Requested resource doesn't exist"}

real	0m0.041s
user	0m0.026s
sys	0m0.011s
```
How can we improve that ?